packages/next/server/base-http.ts

```
@@ -0,0 +1,287 @@
+  basePath: string | undefined
+  i18n?: I18NConfig
+  trailingSlash?: boolean | undefined
+}
+
+export abstract class BaseNextRequest<Body = any> {
+  protected _cookies: NextApiRequestCookies | undefined
+  public abstract headers: IncomingHttpHeaders
+
+  constructor(public method: string, public url: string, public body: Body) {}
+
+  abstract parseBody(limit: string | number): Promise<any>
+
+  // Utils implemented using the abstract methods above
+
+  public get cookies() {
+    if (this._cookies) return this._cookies
+    return (this._cookies = getCookieParser(this.headers)())
+  }
+}
+
+export class NodeNextRequest extends BaseNextRequest<Readable> {
+  public headers = this._req.headers;
+
+  [NEXT_REQUEST_META]: RequestMeta
+
+  get req() {
+    // Need to mimic these changes to the original req object for places where we use it:
+    // render.tsx, api/ssg requests
+    this._req[NEXT_REQUEST_META] = this[NEXT_REQUEST_META]
+    this._req.url = this.url
+    this._req.cookies = this.cookies
+    return this._req
+  }
+
+  constructor(
+    private _req: IncomingMessage & {
+      [NEXT_REQUEST_META]?: RequestMeta
+      cookies?: NextApiRequestCookies
+    }
+  ) {
+    super(_req.method!.toUpperCase(), _req.url!, _req)
+  }
+
+  async parseBody(limit: string | number): Promise<any> {
+    return parseBody(this._req, limit)
+  }
+}
+
+export class WebNextRequest extends BaseNextRequest<ReadableStream | null> {
+  public request: Request
+  public headers: IncomingHttpHeaders
+
+  constructor(request: Request) {
+    const url = new URL(request.url)
+
+    super(
+      request.method,
+      url.href.slice(url.origin.length),
+      request.clone().body
+    )
+    this.request = request
+
+    this.headers = {}
+    for (const [name, value] of request.headers.entries()) {
+      this.headers[name] = value
+    }
+  }
+
+  async parseBody(_limit: string | number): Promise<any> {
+    // TODO: implement parseBody for web
```

Should we throw here to avoid unexpected bugs?